File: src/frontend/api/tesco/type/SearchProduct.ts
export interface SearchProductNutrient {
    name: string,
    valuePer100g: number,
    valuePerServing: number,
}

export interface SearchProductNutrition {
    per100gHeader: string,
    perServingHeader: string,
    nutrients: SearchProductNutrient[],
}

export interface SearchProductGuidelineDailyAmounts {
    headers: Array<string>;
    footers: Array<string>;
    amounts: Array<SearchProductGuidelineDailyAmountsAmount>;
}

export interface SearchProductGuidelineDailyAmountsAmount {
    name: string;
    values: Array<string>;
    percent: string;
    rating: 'LOW' | 'MEDIUM' | 'HIGH';
}

export interface SearchProduct {
    id: string,
    name: string,
    price: number,
    imageUrl: string,
    description: string,
    department: string,
    superDepartment: string,

    barcodeId?: string,
    ingredients?: string[],
    healthScore?: number,
    nutrition?: SearchProductNutrition,
    guidelineDailyAmounts: SearchProductGuidelineDailyAmounts;
}

File: src/backend/src/main/java/com/wjbaker/gocart_api/tesco/type/SearchProduct.java
package com.wjbaker.gocart_api.tesco.type;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

@Builder
@Getter
@Setter
public class SearchProduct {

    private String id;
    private String name;
    private float price;
    private String imageUrl;
    private String description;
    private String department;
    private String superDepartment;
    private String barcodeId;
    private List<String> ingredients;
    private Integer healthScore;
    private Nutrition nutrition;
    private GuidelineDailyAmounts guidelineDailyAmounts;

    @Builder
    @Getter
    @Setter
    public static class Nutrition {

        private String per100gHeader;
        private String perServingHeader;
        private List<Nutrient> nutrients;

        @Builder
        @Getter
        @Setter
        public static class Nutrient {

            private String name;
            private float valuePer100g;
            private float valuePerServing;
        }
    }

    @Getter
    @Setter
    public static class GuidelineDailyAmounts {

        private List<String> headers;
        private List<String> footers;
        private List<GdaAmount> amounts;

        @Getter
        @Setter
        public static class GdaAmount {

            private String name;
            private List<String> values;
            private String percent;
            private GdaRating rating;
        }
    }
}
